File: solutions/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # helper function to found nearby land
        def get_neighbours(i, j):
            res = []
            di = [0, 1, -1, 0]
            dj = [-1, 0, 0, 1]
            for idx in range(4):
                new_i = i + di[idx]
                new_j = j + dj[idx]
                if 0 <= new_i < len(grid[0]) and 0 <= new_j < len(grid):
                    if grid[new_j][new_i] == "1":
                        res.append((new_i, new_j))
            return res

        # turn every island to water to dont count visited ones
        def dfs(i, j):
            if grid[j][i] == "0":
                return
            grid[j][i] = "0"
            neighbours = get_neighbours(i, j)
            for x, y in neighbours:
                if grid[y][x] == "0":
                    continue
                dfs(x, y)
            

        # go through every cell to find land ad count them
        count = 0
        for j in range(len(grid)):
            for i in range(len(grid[0])):
                if grid[j][i] == "1":
                    dfs(i, j)
                    count += 1
        return count

File: solutions/200.go
package main

import "fmt"

func getNeighbours(grid [][]byte, i, j int) [][]int{
	result := [][]int{}
	di := []int{-1, 0, 0, 1}
	dj := []int{0, -1, 1, 0}
	for idx := range di {
		new_i := i + di[idx]
		new_j := j + dj[idx]
		if 0 <= new_i && new_i < len(grid[0]) && 0 <= new_j && new_j < len(grid) {
			if grid[new_j][new_i] == '1' {
				result = append(result, []int{new_i, new_j})
			}
		}
	}
	return result
}

func clearIsland(grid [][]byte, i, j int) {
	if grid[j][i] == '0' {
		return
	}
	grid[j][i] = '0' // чтобы не посещать те же клетки два раза
	neighbours := getNeighbours(grid, i, j)
	for _, neighbour := range neighbours {
		neighbour_i := neighbour[0]
		neighbour_j := neighbour[1]
		if grid[neighbour_j][neighbour_i] == '0' {
			continue
		}
		clearIsland(grid, neighbour_i, neighbour_j)
	}
}
// когда находим сушу, то мы исследуем весь остров и его заменяем на '0'.
func numIslands(grid [][]byte) int {
	result := 0
    for j := range grid {
		for i := range grid[0] {
			if grid[j][i] == '1' {
				clearIsland(grid, i, j)
				result++
			}
		}
	}
	return result
}


func main() {
	grid := [][]byte{
		{'1', '1', '1', '1', '0'},
		{'1', '1', '0', '1', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '0', '0', '0'},
	}
	fmt.Println(numIslands(grid))
}